### PR TITLE
remote-build: initial Windows support

### PR DESCRIPTION
**Pull request by cjp256**
Requires: https://github.com/snapcore/snapcraft/pull/2759

Tested against: https://github.com/cjp256/figlet/tree/use-install

### CODE REVIEW

File: snapcraft/file_utils.py [hunk 1]
```diff
@@ -264,6 +264,9 @@ def create_similar_directory(source: str, destination: str) -> None:
     gid = stat.st_gid
     os.makedirs(destination, exist_ok=True)
 
+    if sys.platform == "win32":
```
Comment (sergiusens) [inline snapcraft/file_utils.py]: Might be a good idea to add the commit description or similar as a code comment

File: snapcraft/internal/sources/__init__.py [hunk 2]
```diff
@@ -111,7 +111,12 @@
         "snap": Snap,
         "": Local,
     }
+elif sys.platform == "win32":
+    from ._git import Git  # noqa
+    from ._local import Local  # noqa
+    from ._tar import Tar  # noqa
```
Comment (sergiusens) [inline snapcraft/internal/sources/__init__.py]: Tar is missing from the list in the commit message

File: snapcraft/file_utils.py [hunk 3]
```diff
@@ -417,3 +418,15 @@ def get_resolved_relative_path(relative_path: str, base_directory: str) -> str:
     filename_relpath = os.path.relpath(filename_abspath, base_directory)
 
     return filename_relpath
+
+
+def _remove_readonly(func, path, excinfo):
+    """Try setting file to writeable if error occurs.
```
Comment (sergiusens) [inline snapcraft/file_utils.py]: this docstring does not follow the PEP recommendation we follow https://www.python.org/dev/peps/pep-0257/#one-line-docstrings
Is this a code comment in disguise?

File: snapcraft/file_utils.py [hunk 4]
```diff
@@ -417,3 +418,15 @@ def get_resolved_relative_path(relative_path: str, base_directory: str) -> str:
...
+    Known to be required on Windows."""
+    os.chmod(path, stat.S_IWRITE)
+    func(path)
+
+
+def rmtree(path: str) -> None:
+    """Cross-platform rmtree implementation."""
+    shutil.rmtree(path, onerror=_remove_readonly)
```
Comment (sergiusens) [inline snapcraft/file_utils.py]: please add an newline to get github to get rid of that ugly symbol :laughing:

File: snapcraft/file_utils.py [hunk 5]
```diff
@@ -429,4 +430,17 @@ def _remove_readonly(func, path, excinfo):
 
 def rmtree(path: str) -> None:
     """Cross-platform rmtree implementation."""
-    shutil.rmtree(path, onerror=_remove_readonly)
\ No newline at end of file
+    shutil.rmtree(path, onerror=_remove_readonly)
+
+
+def format_linux_path(path: str) -> str:
+    """Convert native OS path to linux path.
```
Comment (sergiusens) [inline snapcraft/file_utils.py]: please follow https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

File: snapcraft/file_utils.py [hunk 6]
```diff
@@ -429,4 +430,17 @@ def _remove_readonly(func, path, excinfo):
...
+    Note that this is really only useful for relative paths...
+    e.g. "some\\path\\here" -> "some/path/here"
+    """
+
+    if sys.platform == "win32":
+        parts = pathlib.WindowsPath(path).parts
```
Comment (sergiusens) [inline snapcraft/file_utils.py]: why not pathlib.PureWindowsPath(...).to_posix()?

File: snapcraft/internal/remote_build/_worktree.py [hunk 7]
```diff
@@ -145,7 +145,10 @@ def _archive_part_sources(self, part_name: str, selector=None):
         os.makedirs(os.path.split(archive_path)[0], exist_ok=True)
         command = ["tar", "czf", archive_path, "-C", source_path, "."]
         subprocess.check_call(command)
-        return os.path.relpath(archive_path, self._repo_dir)
+        relpath = os.path.relpath(archive_path, self._repo_dir)
```
Comment (sergiusens) [inline snapcraft/internal/remote_build/_worktree.py]: you could make pathlib.Path objects out of this and return .to_posix ... Path will work appropriately depending on platform

File: snapcraft/file_utils.py (hunk 6 above)
Comment (cjp256) [reply snapcraft/file_utils.py]: done.

File: snapcraft/file_utils.py (hunk 5 above)
Comment (cjp256) [reply snapcraft/file_utils.py]: removed function.

File: snapcraft/internal/remote_build/_worktree.py (hunk 7 above)
Comment (cjp256) [reply snapcraft/internal/remote_build/_worktree.py]: done (as_posix)

File: snapcraft/file_utils.py (hunk 4 above)
Comment (cjp256) [reply snapcraft/file_utils.py]: done!

File: snapcraft/file_utils.py (hunk 3 above)
Comment (cjp256) [reply snapcraft/file_utils.py]: fixed (Code comment)

File: snapcraft/internal/sources/__init__.py (hunk 2 above)
Comment (cjp256) [reply snapcraft/internal/sources/__init__.py]: Fixed!

File: snapcraft/file_utils.py (hunk 1 above)
Comment (cjp256) [reply snapcraft/file_utils.py]: Good call, fixed!

**Pull request merged**